plugins/linux-v4l2/v4l2-input.c

```diff
@@ -406,26 +411,77 @@ static void v4l2_device_list(obs_property_t *prop, obs_data_t *settings)
 #endif
 
 		if (!(caps & V4L2_CAP_VIDEO_CAPTURE)) {
-			blog(LOG_INFO, "%s seems to not support video capture",
+			blog(LOG_WARNING,
+			     "%s seems to not support video capture",
 			     device.array);
-			v4l2_close(fd);
 			continue;
 		}
 
-		/* make sure device names are unique */
-		char unique_device_name[68];
-		sprintf(unique_device_name, "%s (%s)", video_cap.card,
-			video_cap.bus_info);
-		obs_property_list_add_string(prop, unique_device_name,
-					     device.array);
-		blog(LOG_INFO, "Found device '%s' at %s", video_cap.card,
-		     device.array);
-
-		/* check if this is the currently used device */
-		if (cur_device_name && !strcmp(cur_device_name, device.array))
-			cur_device_found = true;
+		bool device_already_added = false;
+
+		if (!allow_duplicates) {
+			size_t listidx = 0;
+			const char *item_name;
+			while (NULL != (item_name = obs_property_list_item_name(
+						prop, listidx++))) {
+				if (NULL !=
+				    strstr(item_name, video_cap.bus_info)) {
+					device_already_added = true;
+					break;
+				}
+			}
+		}
 
-		v4l2_close(fd);
+		if (!device_already_added) {
+			char unique_device_name[68];
+			if (0 == strcmp("/dev/v4l/by-path/", basedir)) {
+				sprintf(unique_device_name, "%s (%s)",
+					video_cap.bus_info, video_cap.card);
+			} else {
+				sprintf(unique_device_name, "%s (%s)",
+					video_cap.card, video_cap.bus_info);
+			}
+			obs_property_list_add_string(prop, unique_device_name,
+						     device.array);
+			blog(LOG_INFO, "Found device '%s' at %s",
+			     video_cap.card, device.array);
+		}
+	}
+
+	closedir(dirp);
+	dstr_free(&device);
+}
+
+/*
+ * List available devices
+ */
+static void v4l2_device_list(obs_property_t *prop, obs_data_t *settings)
+{
+	bool cur_device_found;
+	size_t cur_device_index;
+	const char *cur_device_name;
+
+	obs_property_list_clear(prop);
+
+#ifdef __FreeBSD__
```

Comment:
With this structure we should be able to clear all the freebsd ifdefs inside of the new function it seems? But maybe im missing something.

Reply:
FreeBSD enumerates all files from /dev which contains "video".
In case of linux we're enumarating all files from /sys/class/video4linux to get the file names. However we're opening device files from /dev.
It's how it used to work before my changes.